processing/src/main/java/io/druid/segment/IndexIO.java

```diff
@@ -361,28 +450,48 @@ public static void validateTwoSegments(File dir1, File dir2) throws IOException
     {
       final QueryableIndexIndexableAdapter adapter1 = new QueryableIndexIndexableAdapter(loadIndex(dir1));
       final QueryableIndexIndexableAdapter adapter2 = new QueryableIndexIndexableAdapter(loadIndex(dir2));
-      if(adapter1.getNumRows() != adapter2.getNumRows()){
-        throw new IOException("Validation failure - Row count mismatch");
+      if (adapter1.getNumRows() != adapter2.getNumRows()) {
+        throw new SegmentValidationException("Row count mismatch");
+      }
+      {
+        final Set<String> dimNames1 = Sets.newHashSet(adapter1.getDimensionNames());
+        final Set<String> dimNames2 = Sets.newHashSet(adapter2.getDimensionNames());
+        if (!dimNames1.equals(dimNames2)) {
+          throw new SegmentValidationException("Dimension names differ %s vs %s", dimNames1, dimNames2);
+        }
+        final Set<String> metNames1 = Sets.newHashSet(adapter1.getMetricNames());
+        final Set<String> metNames2 = Sets.newHashSet(adapter2.getMetricNames());
+        if (!metNames1.equals(metNames2)) {
+          throw new SegmentValidationException("Metric names differ %s vs %s", metNames1, metNames2);
+        }
       }
       final Iterator<Rowboat> it1 = adapter1.getRows().iterator();
       final Iterator<Rowboat> it2 = adapter2.getRows().iterator();
       long row = 0L;
-      while(it1.hasNext()){
-        if(it1.hasNext() ^ it2.hasNext()){
-          throw new IOException("Validation failure - Iterator doesn't have enough");
+      while (it1.hasNext()) {
+        if (it1.hasNext() ^ it2.hasNext()) {
+          throw new SegmentValidationException("Iterator doesn't have enough");
         }
         final Rowboat rb1 = it1.next();
         final Rowboat rb2 = it2.next();
         ++row;
-        if(rb1.compareTo(rb2) != 0){
-          throw new IOException(String.format("Validation failure on row %d: [%s] vs [%s]", row, rb1, rb2));
+        if (rb1.getRowNum() != rb2.getRowNum()) {
```

Comment:
ok, I was just asking if there was a reason we added it, since we didn't have it before